### ISSUE TITLE
Automatic hyperlinks

### ISSUE THREAD

**Issue by anseljh** [commonform/commonform-docx#1]
Sometimes I have URLs in my contracts. It would be nice if this package would detect them and create hyperlink objects in the output Word file.

I think the right approach is to keep written-out URLs in the text, and just wrap them with hyperlink markup.

Some possbily helpful info on how to accomplish this:
- [WordprocessingML Basics](http://download.microsoft.com/download/5/1/6/5160beb0-ed51-4fa5-9e73-334114e7fb24/WordprocessingMLBasics.pptx) presentation from Microsoft
- [ECMA-376](http://www.ecma-international.org/publications/standards/Ecma-376.htm) Part 1, Section 15.3 (etc.)

**Comment by kemitchell** [comment]
I see two parts here:
1. Detect when URLs appear in fully resolved forms. It's all strings at that point, whether the URL was written into a form or placed as a fill-in-the-blank value.
2. Output the correct XML.

A few leads:

https://www.npmjs.com/package/get-urls

https://github.com/commonform/commonform-docx/blob/master/templates/run.js#L47

**Comment by anseljh** [comment]
Note to self, for later:

``` xml
<w:hyperlink r:id="rId7" w:history="1"><w:r w:rsidRPr="009B49CC"><w:rPr><w:rStyle w:val="Hyperlink"/></w:rPr><w:t>Link content</w:t></w:r></w:hyperlink>
```

**Comment by anseljh** [comment]
More links from Kyle:
- [Anatomy of a WordProcessingML File](http://officeopenxml.com/anatomyofOOXML.php)
- [Open XML SDK 2.5 for Microsoft Office](https://www.microsoft.com/en-us/download/details.aspx?id=30425)

**Comment by kemitchell** [comment]
This looks promising: [url-regex](https://github.com/kevva/url-regex)

**Comment by kemitchell** [comment]
http://officeopenxml.com/WPhyperlink.php

**Comment by anseljh** [comment]
Cool, thanks. I may have some time over the weekend to put this together.

**Comment by kemitchell** [comment]
Be warned: DOCX makes this kind of subtle. A nice first cut might be:
1. hyperlink a fill-in-the-blank value if it's a valid URI
2. hyperlink URI substrings of plain text

I personally wouldn't be too worried about links in headings, terms, or their combination with plain text.

**Comment by anseljh** [comment]
Right, makes sense. I'm looking forward to this because I put links in my
website terms forms, and have to go linkify them manually after building
the .DOCX files. Obviously this will be great for HTML rendering too (and
presumably much easier).

**Comment by kemitchell** [comment]
This came up again, and I looked at it again.

Basically, we can do the hyperlinks in the document easily enough. But we have to put the URLs they point do in a different XML file within the .docx ZIP archive.

document.xml.rels needs something like this:

```xml
<Relationship Id="example" Type="http://. . ./hyperlink" Target="http://www.example.com/" TargetMode="External"/>
```

in order to do this in document.xml:

```
<w:hyperlink r:id="example">
<w:r>
<w:rPr>
<w:rStyle w:val="Hyperlink"/>
</w:rPr>
<w:t>example.com</w:t>
</w:r>
</w:hyperlink>
```
